### PR TITLE
Crossbar module for allotments

### PR DESCRIPTION
**Pull request by Dinkor**
Show current allotments configuration.
Update allotments configuration (only reseller/admin).
Show consumed allotments in current cycle.

### CODE REVIEW

File: applications/crossbar/src/modules/cb_allotments.erl [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+-spec load_consumed(cb_context:context(), api_binary(), wh_json:objects()) -> cb_context:context().
+load_consumed(Context, MoDb, Allotments) ->
+    {_, Result} = wh_json:foldl(fun foldl_consumed/3, {MoDb, []}, Allotments),
+    cb_context:setters(Context
+                       ,[{fun cb_context:set_resp_status/2, 'success'}
+                         ,{fun cb_context:set_resp_data/2, Result}
+                        ]).
+
+-spec foldl_consumed(api_binary(), wh_json:object(), {cb_context:context(), api_binary(), wh_json:objects()}) -> wh_json:objects().
+foldl_consumed(Classification, Value, {MoDb, Acc}) ->
```
Comment (lazedo) [inline applications/crossbar/src/modules/cb_allotments.erl]: @Dinkor spec seems wrong. declares tuple/3 in 3rd argument but function handles tuple/2

Comment (Dinkor) [inline applications/crossbar/src/modules/cb_allotments.erl]: Yes. Changed function, but forget about spec.

File: applications/crossbar/src/modules/cb_allotments.erl [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+-spec load_consumed(cb_context:context()) -> cb_context:context().
+load_consumed(Context) ->
+    Context1 = crossbar_doc:load(?PVT_TYPE, Context),
+    Allotments = wh_json:get_json_value(?PVT_ALLOTMENTS, cb_context:doc(Context1), wh_json:new()),
+    load_consumed(Context, Allotments).
+
+-spec load_consumed(cb_context:context(), wh_json:objects()) -> cb_context:context().
+load_consumed(Context, Allotments) ->
+    ReqCycleStart = cb_context:req_param(Context, <<"start">>),
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_allotments.erl]: I'd like to see `from` and `to` used, to mirror `created_from/created_to` and `modified_from/modified_to`. Even better, you could use `consumed_from/consumed_to` and look at adapting to use `cb_modules_util:range_view_options` and `cb_modules_util:range_modb_view_options` to get the list of MODbs needed to fulfill the valid range specified.

File: applications/crossbar/src/modules/cb_allotments.erl [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+    DateTime = calendar:universal_time(),
+    MoDb = wh_util:format_account_mod_id(cb_context:account_id(Context), calendar:datetime_to_gregorian_seconds(DateTime)),
+    {_, _, Result} = wh_json:foldl(fun foldl_consumed_datetime/3, {DateTime, MoDb, []}, Allotments),
+    cb_context:setters(Context
+                       ,[{fun cb_context:set_resp_status/2, 'success'}
+                         ,{fun cb_context:set_resp_data/2, Result}
+                        ]);
+
+load_consumed(Context, Allotments, ReqStart, ReqEnd) ->
+    case normalize_timestamps({ReqStart, ReqEnd}) of
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_allotments.erl]: I'd like to see validation errors (see `cb_context:add_validation_error/4`) used here for both timestamps. Meaning, if both are invalid, two validation errors should be in the response.

File: applications/crossbar/doc/allotments.md [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+           ]
+       }
+```
+So if we already have counsumed calls:
+Class1 - 300
+Class2 - 60
+Class3 - 180
+
+As result next call wil have this free seconds:
+Class1 - 60 (300 Class1 + 60 Class2 + 180 Class = 540, 600-540 = 60)
```
Comment (jamesaimonetti) [inline applications/crossbar/doc/allotments.md]: Missing a 3 in the last Class

File: applications/crossbar/src/modules/cb_allotments.erl [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+%%--------------------------------------------------------------------
+-spec update_allotments(cb_context:context()) -> cb_context:context().
+update_allotments(Context) ->
+    Doc = cb_context:doc(Context),
+    Allotments = cb_context:req_data(Context),
+    NewDoc = wh_json:set_value(?PVT_ALLOTMENTS, Allotments, Doc),
+    Context1 = crossbar_doc:save(cb_context:set_doc(Context, NewDoc)),
+    cb_context:set_resp_data(Context1, Allotments).
+
+-spec cycle_start(ne_binary(), wh_datetime()) -> gregorian_seconds().
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_allotments.erl]: This all looks similar to code in `j5_allotments`; let's break this out into its own core file `core/kazoo_documents/src/kzd_allotment.erl` The building of the allotment_consumption docs should be moved here too (but that can be filed as a ticket - not necessary for this PR).

**Comment by Dinkor** [comment]
Rebase branch to current master

**Comment by Dinkor** [comment]
Rebase is intereference with commit 6da52f4831e23652479dbc9e523f500bfbba8652 in applications/crossbar/src/modules/cb_modules_util.erl.
Both commits create functions with name create_from/create_to

**Comment by fenollp** [comment]
@Dinkor You'll have to resolve the conflicts yourself :)

**Comment by Dinkor** [comment]
Of course, I'll do it.
Just did not have time because of [this](https://groups.google.com/forum/#!topic/2600hz-dev/_AkVLZHbzxk).

**Comment by fenollp** [comment]
Wow! Nice catch

File: applications/crossbar/doc/allotments.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
+/*
+Section: Crossbar
+Title: Allotments
+Language: en-US
+*/
```
Comment (jamesaimonetti) [inline applications/crossbar/doc/allotments.md]: Can you add `Version: 3.22` so the reader knows what version this doc applies to? Big thanks for doing the doc!

File: applications/crossbar/src/modules/cb_allotments.erl [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    case wh_util:is_empty(T) of
+        'true' -> 'undefined';
+        'false' -> wh_util:to_integer(T)
+    end.
+
+-spec on_successful_validation(cb_context:context()) -> cb_context:context().
+on_successful_validation(Context) ->
+    case is_allowed(Context) of
+        'true' -> maybe_handle_load_failure(crossbar_doc:load(?PVT_TYPE, Context));
+        'false' -> crossbar_util:response_400("sub-accounts of non-master resellers must contact the reseller to change their allotments", wh_json:new(), Context)
```
Comment (jamesaimonetti) [inline applications/crossbar/src/modules/cb_allotments.erl]: The first arg should be a binary, not a string.

**Comment by dschreiber** [comment]
Why was this not accepted?

**Comment by lazedo** [comment]
@dschreiber it wasn't ? i see @k-anderson merged the commit

**Comment by fenollp** [comment]
@dschreiber Note PR was squashed before merge meaning sha1s of this PR do not appear in history, but [merge commit does](https://github.com/2600hz/kazoo/commit/ad2ab44a1febfe1f8755821c910cb48c913f1e45)

**Pull request merged**